### PR TITLE
basic C++20 support

### PR DESCRIPTION
**Pull request by andrewcorrigan**
# Summary

- This PR is a feature
- It does the following:
  - adds a c++20 option
  - updates `BLT` a few commits ahead to [e35f490a8a8b1689e99b5f4308b5251f97eb36cf](https://github.com/LLNL/blt/commit/e35f490a8a8b1689e99b5f4308b5251f97eb36cf) for C++20 support
  - prefers `std::invoke_result` over c++17-deprecated/c++20-removed functionality `std::result_of` [ [cppreference] ](https://en.cppreference.com/w/cpp/types/result_of)

### CODE REVIEW

**Review by trws — APPROVED** [review]
Looks good to me, as far as I know so far there shouldn't be anything else we *need* to do for C++20 support.  There are probably lots of opportunities there, but we'll have to give some thought to how and if we can take advantage of them.

File: include/RAJA/internal/foldl.hpp [hunk 1]
```diff
@@ -48,6 +48,28 @@ struct foldl_impl<Op, Arg1> {
   using Ret = Arg1;
 };
 
+#if defined(__cpp_lib_is_invocable) && (__cpp_lib_is_invocable >= 201703L)
```
Comment (rhornung67) [inline include/RAJA/internal/foldl.hpp:51]: @trws should we be checking this sort of thing here, following the same pattern as for other language features? https://github.com/LLNL/RAJA/blob/develop/include/RAJA/config.hpp.in#L43

Comment (trws) [reply include/RAJA/internal/foldl.hpp:51]: Normally I'd say yes, but for something we only need to check in one place, I think it's fine for it to be inline like this.  If it needs to be managed for compiler differences or gets checked elsewhere then we should move it to config.

**Review by rhornung67 — APPROVED** [review]
Thanks @andrewcorrigan. In order to merge this, we need to pull your forked branch into our repo and make a new PR. Are you OK with this?

**Review by artv3 — APPROVED** [review]

**Comment by andrewcorrigan** [comment]
> Thanks @andrewcorrigan. In order to merge this, we need to pull your forked branch into our repo and make a new PR. Are you OK with this?

I'm OK with it.

For my understanding, do you need me to do something differently next time I submit a PR?

File: include/RAJA/internal/foldl.hpp (hunk 1 above)
Comment (mdavis36) [reply include/RAJA/internal/foldl.hpp:51]: I wouldn't mind seeing it moved to config, it's nice to have a list of all the compiler feature we may or may not use in one place.

Comment (andrewcorrigan) [reply include/RAJA/internal/foldl.hpp:51]: I'd be happy to revise the PR to define a macro in config if that's the consensus.

Comment (trws) [reply include/RAJA/internal/foldl.hpp:51]: Yeah, might as well.

**Comment by trws** [comment]
> > Thanks @andrewcorrigan. In order to merge this, we need to pull your forked branch into our repo and make a new PR. Are you OK with this?
> 
> I'm OK with it.
> 
> For my understanding, do you need me to do something differently next time I submit a PR?

If you contribute regularly we might give you access to do more, but for now it's a limitation of our CI setup that we don't have a way to activate the gitlab CI for PRs from forks.  It's on the list of features we've requested, but until we have a way to do it with a comment or similar we have to do this two-step dance to get all the tests run.

File: include/RAJA/internal/foldl.hpp (hunk 1 above)
Comment (andrewcorrigan) [reply include/RAJA/internal/foldl.hpp]: done.   please let me know if further changes are required.

**Comment by rhornung67** [comment]
@andrewcorrigan I've been away for a bit, I'll get this pulled in to our repo soon and tag you on the new PR.

**Comment by rhornung67** [comment]
@andrewcorrigan here is a link to the PR that will get merged based on this: https://github.com/LLNL/RAJA/pull/1240

**Comment by rhornung67** [comment]
@andrewcorrigan I switched us over to new queues for Gitlab CI. This should go through much faster now.

**Comment by rhornung67** [comment]
Closing this one. https://github.com/LLNL/RAJA/pull/1240 is merged.

**Pull request closed without merging**